discovery_child_development/notebooks/01_openalex_clustering.py

```diff
@@ -0,0 +1,262 @@
+# * Find the top terms per cluster
+
+# +
+import boto3
+from io import BytesIO
+import os
+from dotenv import find_dotenv, load_dotenv
+from time import time
+import pandas as pd
+import numpy as np
+import pickle
+
+import umap.umap_ as umap
+import hdbscan
+import altair as alt
+
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.preprocessing import normalize
+
+# -
+
+# We store the concepts we're using as a variable so that they can be pasted into file names (the metaflow process stores the concepts that were used to get the data in the file name, and it seems wise to preserve this information at different stages of the process).
+#
+# At some point it may help to store these in a config file, in case we want to work with different sets of concepts at the same time for comparison.
+
+CONCEPTS = "C109260823|C2993937534|C2777082460|C2911196330|C2993037610|C2779415726|C2781192327|C15471489|C178229462"
+
+# +
+load_dotenv()
+
+S3_BUCKET = os.environ["S3_BUCKET"]
+
+s3_client = boto3.client("s3")
+# -
+
+# Read in the text data
+text_data_path = f"inputs/data/openAlex/openalex_abstracts_{CONCEPTS}_year-2023.csv"
+response = s3_client.get_object(Bucket=S3_BUCKET, Key=text_data_path)
+csv_data = response["Body"].read()
+openalex_text_df = pd.read_csv(BytesIO(csv_data), index_col=0)
+openalex_text_df.head()
+
+# Put the text in a format suitable for embedding, clustering etc
+openalex_docs = openalex_text_df["text"].tolist()
+
+# get rid of NaNs
+print(f"Number of docs before removing NaNs: {len(openalex_docs)}")
+openalex_docs = [doc for doc in openalex_docs if isinstance(doc, str)]
+print(f"Number of docs after removing NaNs: {len(openalex_docs)}")
+
+# apply the same filtering on the dataframe
+print(f"Number of docs in dataframe before removing NaNs: {len(openalex_text_df)}")
+openalex_text_df = openalex_text_df[
+    openalex_text_df["text"].apply(lambda x: isinstance(x, str))
+]
+print(f"Number of docs in dataframe AFTER removing NaNs: {len(openalex_text_df)}")
+
+# The embeddings were computed separately because this part can be time consuming. We can simply load them from the pickle file:
+
+# +
+# Load embeddings
+buffer = BytesIO()
+s3_client.download_fileobj(
+    S3_BUCKET, "inputs/data/openAlex/vectors/sentence_vectors_384.pkl", buffer
+)
+buffer.seek(0)
+openalex_vectors = pickle.load(buffer)
+
+# Normalize the vectors
+openalex_vectors_normalized = normalize(openalex_vectors)
+# -
+
+openalex_vectors_normalized.shape
+
+# # TFIDF
+#
+# We first try simple TFIDF vectors before using sentence embeddings.
+
+tfidf_vectorizer = TfidfVectorizer(
+    input="content", max_df=0.5, min_df=1, stop_words="english"
+)
+
+t0 = time()
+tfidf_vectors = tfidf_vectorizer.fit_transform(openalex_docs)
+print(f"vectorization done in {time() - t0:.3f} s")
+print(f"n_samples: {tfidf_vectors.shape[0]}, n_features: {tfidf_vectors.shape[1]}")
+
+
+def apply_umap_and_cluster(vectors, seed=None):
```

Comment:
Is it ok to refactor these functions on a separate PR (if they stay in use)? Because there are other changes that should be made if they stay in use, eg I think the 2d dimension reduction should be a separate function (like how you have it in `embeddings_utils.py` in the other Innovation Sweet Spots repo

Reply:
Yes, sounds good!